### PR TITLE
User Service Test setup

### LINKED ISSUE #206: UserService class testing

### PR DESCRIPTION
**Pull request by ralphgregorio**
Resolves #206 

# Description

Fixed our dependencies regarding testing.

Also finished a thorough test of UserService

Also fixed pipeline not working correctly, and backend dependencies

Also improved pipeline by only running jobs if only related files are changed for pull_requests

Learning resource for testing:
https://www.freecodecamp.org/news/unit-testing-services-endpoints-and-repositories-in-spring-boot-4b7d9dc2b772/

### CODE REVIEW

File: backend/src/test/java/service/UserServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,535 @@
...
+        User user = mockUser();
+        UserDto userDto = mockUserDto();
+
+        JSONObject expected = new JSONObject();
+
+        expected.put(BeapEngineConstants.SUCCESS_STR, true);
+        expected.put("userDto", userDto);
+        expected.put("status_code", HttpStatus.CREATED.value());
+
+        when(userDao.findByUsername(user.getUsername())).thenReturn(null);
```
Comment (redwing14e) [inline backend/src/test/java/service/UserServiceTest.java:115]: should be mocking UserService.java line 74
`userMapper.dto2Model(userDto, new User());`

File: backend/src/test/java/service/UserServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,535 @@
...
+        UserDto userDto = mockUserDto();
+
+        JSONObject expected = new JSONObject();
+
+        expected.put(BeapEngineConstants.SUCCESS_STR, false);
+        expected.put("message", "There is already another account with username: " + userDto.getUsername());
+        expected.put("status_code", HttpStatus.INTERNAL_SERVER_ERROR.value());
+
+        when(userDao.findByUsername(user.getUsername())).thenReturn(user);
+        when(userDao.save(Mockito.any(User.class))).thenReturn(1L);
```
Comment (redwing14e) [inline backend/src/test/java/service/UserServiceTest.java]: I believe this is not called

File: backend/src/test/java/service/UserServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+    @Test
+    /*
+     * Preconditions: User exists in the database
+     * Post-conditions: Returns UserDto object
+     */
+    public void testGet() {
+        User user = mockUser();
+        UserDto expected = mockUserDto();
+
+        when(userDao.get(anyLong())).thenReturn(user);
```
Comment (redwing14e) [inline backend/src/test/java/service/UserServiceTest.java]: if possible switch this to check that the correct long is being sent not anyLong()

File: backend/src/test/java/service/UserServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,535 @@
...
+        expected.put(BeapEngineConstants.SUCCESS_STR, true);
+        expected.put("status_code", HttpStatus.OK.value());
+        expected.put("Authorities", userDetails.getAuthorities());
+        expected.put("userDetails", userDetails);
+        expected.put("userDto", userDto);
+
+        when(userDao.findByUserPass(user.getUsername(), user.getPassword())).thenReturn(user);
+        when(userMapper.model2Dto(Mockito.<User>anyObject(), Mockito.any(UserDto.class))).thenReturn(userDto);
+        when(userMapper.dto2Model(Mockito.<UserDto>anyObject(), Mockito.any(User.class))).thenReturn(user);
+        when(incorrectLoginsService.getByUserId(Mockito.anyLong())).thenReturn(
```
Comment (redwing14e) [inline backend/src/test/java/service/UserServiceTest.java]: again is it possible to check if these are getting called with the correct values

File: backend/src/test/java/service/UserServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+    /*
+     * Preconditions: User exists in the database
+     * Post-conditions: Returns UserDto object
+     */
+    public void testGet() {
+        User user = mockUser();
+        UserDto expected = mockUserDto();
+
+        when(userDao.get(anyLong())).thenReturn(user);
+        when(userMapper.model2Dto(Mockito.<User>anyObject(), Mockito.<UserDto>anyObject())).thenReturn(expected);
```
Comment (redwing14e) [inline backend/src/test/java/service/UserServiceTest.java]: can Mockito.<User>anyObject() be switched to `user` (that was made on line 163)

File: backend/src/test/java/service/UserServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,535 @@
...
+        when(userMapper.model2Dto(Mockito.<User>anyObject(), Mockito.any(UserDto.class))).thenReturn(userDto);
+        when(userMapper.dto2Model(Mockito.<UserDto>anyObject(), Mockito.any(User.class))).thenReturn(user);
+        when(incorrectLoginsService.getByUserId(Mockito.anyLong())).thenReturn(
+                mockIncorrectLoginsDto(
+                        userDto,
+                        false,
+                        0,
+                        getTimestamp(2023, 0, 1, 0, 0)
+                )
+        );
```
Comment (redwing14e) [inline backend/src/test/java/service/UserServiceTest.java]: based on some of the comments in user service I believe this should be returning null when there has not ever been an incorrect login attempt. but also I can see that this tests the case where there were incorrect login attempts but they are not enough to lock the user out (if this is what you are testing then line 222 should not be 0).

File: backend/src/test/java/service/UserServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,535 @@
...
+        verify(incorrectLoginsService).update(argument.capture());
+        assertEquals(result, expected);
+        assertEquals(1, argument.getValue().getIncorrectAttempts());
+    }
+
+    @Test
+    /*
+     * Preconditions: Username is invalid and is the first invalid attempt
+     * Post-conditions: Returns JSONObject indicating invalid login attempt and increments incorrect attempt count
+     */
```
Comment (redwing14e) [inline backend/src/test/java/service/UserServiceTest.java]: should probably be for incorrect password not username

File: backend/src/test/java/service/UserServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,535 @@
...
+        verify(incorrectLoginsService).update(argument.capture());
+        assertEquals(results, expected);
+        assertEquals(0, argument.getValue().getIncorrectAttempts());
+        assertFalse(argument.getValue().getLocked());
+        assertNull(argument.getValue().getLockedDate());
+    }
+
+    @Test
+    /*
+     * Preconditions: Username is invalid
```
Comment (redwing14e) [inline backend/src/test/java/service/UserServiceTest.java]: this is actually testing an invalid password.

File: backend/src/test/java/service/UserServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,535 @@
...
+        when(userDao.findByUsername(user.getUsername())).thenReturn(user);
+        when(userMapper.model2Dto(Mockito.<User>anyObject(), Mockito.<UserDto>anyObject())).thenReturn(userDto);
+        when(userMapper.dto2Model(Mockito.<UserDto>anyObject(), Mockito.<User>anyObject())).thenReturn(user);
+
+        when(util.dateDifference(Mockito.<Date>any(), Mockito.<Date>any())).thenReturn(24L);
+        when(incorrectLoginsService.getByUserId(Mockito.anyLong())).thenReturn(
+                mockIncorrectLoginsDto(
+                        userDto,
+                        true,
+                        0,
```
Comment (redwing14e) [inline backend/src/test/java/service/UserServiceTest.java]: would be higher than 0 but does not actually matter for it is not checked

File: backend/src/test/java/service/UserServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,535 @@
...
+
+        when(userDao.findByUserPass(user.getUsername(), user.getPassword())).thenReturn(null);
+        when(userDao.findByUsername(user.getUsername())).thenReturn(user);
+        when(userMapper.model2Dto(Mockito.<User>anyObject(), Mockito.<UserDto>anyObject())).thenReturn(userDto);
+        when(util.dateDifference(Mockito.<Date>any(), Mockito.<Date>any())).thenReturn(2L);
+        when(incorrectLoginsService.getByUserId(Mockito.anyLong())).thenReturn(
+                mockIncorrectLoginsDto(
+                        userDto,
+                        true,
+                        0,
```
Comment (redwing14e) [inline backend/src/test/java/service/UserServiceTest.java]: would be larger than 0 but is not actually checked in this logic.

**Review by redwing14e — CHANGES_REQUESTED** [review]
if possible would like to see more checks on if the correct values are getting called.  
another possible test that could be added is to test when the lockout attempt happens on an invalid login attempt

File: backend/src/test/java/service/UserServiceTest.java (hunk 2 above)
Comment (ralphgregorio) [reply backend/src/test/java/service/UserServiceTest.java]: yep ur right the useDao.save is not called

File: backend/src/test/java/service/UserServiceTest.java (hunk 5 above)
Comment (ralphgregorio) [reply backend/src/test/java/service/UserServiceTest.java]: yep it can!

File: backend/src/test/java/service/UserServiceTest.java (hunk 6 above)
Comment (ralphgregorio) [reply backend/src/test/java/service/UserServiceTest.java]: yep you are right that statement wont even be run in this specific test case

File: backend/src/test/java/service/UserServiceTest.java (hunk 10 above)
Comment (ralphgregorio) [reply backend/src/test/java/service/UserServiceTest.java]: ill change it for better understanding of our testing

**Review by redwing14e — APPROVED** [review]
looks good

**Pull request merged**